### ISSUE TITLE
Performance improvements for values on maps

### ISSUE THREAD

**Issue by marcelgerber** [owid/owid-grapher#4871]
Rendering values on maps, especially external labels, is expensive - and there are some wins to be had.

- *Cluster annotations before running force simulation*
  - Context: The force simulation in `minimiseLabelCollisions` is one of the most expensive steps
  - Before we run it, we can cluster annotations, and run the force simulation on the resulting clusters
  - That's because the labels for Norway and Madagascar are unlikely to ever overlap :)
  - A [suitable clustering algorithm](https://scikit-learn.org/stable/auto_examples/cluster/plot_cluster_comparison.html) would be [DBScan](https://scikit-learn.org/stable/modules/generated/sklearn.cluster.DBSCAN.html)
    - it creates clusters based on a max distance between neighbouring points, and, unlike k-means, we don't need to provide a number of clusters beforehand
    - also, not every point has to go into a cluster, points can also be outliers - for these, we don't need to run the force simulation at all
- *Cut off points outside the viewport*
  - Currently, we disregard the map zoom level, and render annotations for all countries, even when they're completely cut off because of the zoom level
  - We should exclude these annotations, to reduce the number of labels we input into force simulation etc.
  - Relevant link: https://observablehq.com/@d3/testing-projection-visibility
- _Vague_: use *quadtrees* in more places
  - Quadtrees are versatile and fast data structures, and we could probably use them for use cases like the nearest-neighbour search (`getGeoFeaturesWithinRadius`)
  - Relevant link: https://observablehq.com/@d3/quadtree-findincircle

**Comment by sophiamersmann** [comment]
Adding this to our current cycle, as discussed. Thanks again for taking a look, Marcel :)
